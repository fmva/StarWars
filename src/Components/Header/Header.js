import { Component } from 'react';
import React from 'react';
import { Button, Drawer } from 'antd';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import './Header.less';
import Logo from './Img/Star-Wars-Logo.png';

class Header extends Component {
  static propTypes = {
    /** list menu*/
    menu: PropTypes.array,
    /** menu click handler */
    onClickMenu: PropTypes.func,
    /** show/hide drawer menu */
    isDrawerMenu: PropTypes.bool
  };

  constructor(props) {
    super(props);
    this.handlerClickMenu = this.handlerClickMenu.bind(this);
  }

  /**
   * render list of menu
   * @param {Array} arr - array of menu
   * @param {String} className - class name of ul element
   * @return {Object}
   */
  renderMenu(arr, className) {
    const listItems = arr.map((item, key) => (
      <li key={key}>
        <div>
          <Link to={`/${item.path}`}>{item.title}</Link>
        </div>
      </li>
    ));

    return <ul className={className}>{listItems}</ul>;
  }

  /**
   * click drawer menu
   */
  handlerClickMenu() {
    this.props.onClickMenu();
  }

  render() {
    const menu = this.props.menu && this.props.menu.length > 0 ? this.props.menu : [];
    const isDrawerMenu = this.props.isDrawerMenu > 0 ? this.props.isDrawerMenu : false;

    return (
      <header>
        <div className="main-size header-menu">
          {this.renderMenu(menu, 'menu-list phone-visible-none')}

          <div className="icon-bars-block phone-visible-block">
            <i className="icon-bars" title="Menu" onClick={this.handlerClickMenu} />
          </div>
          <img className="sw-logo" src={Logo} alt="StarWars" />
        </div>
        <Drawer
          className="list-drawer"
          placement="left"
          closable={false}
          onClose={this.handlerClickMenu}
          visible={isDrawerMenu}
        >
          <div className="menu-close-block">
            <Button
              className="menu-close-icon"
              type="primary"
              icon="arrow-left"
              onClick={this.handlerClickMenu}
            />
          </div>
          {this.renderMenu(menu, 'menu-list-drawer')}
        </Drawer>
      </header>
    );
  }
}

export default Header;
