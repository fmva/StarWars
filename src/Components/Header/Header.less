@import "../../Less/overall.less";

header {
  background-color: #439cff;
  height:@header-height;
}

.header-menu{
  display: -webkit-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-align-items: flex-start;
  align-items: flex-start;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  @media @phone {
    align-items:center;
  }
}

.sw-logo{
  width: 100px;
  padding: 5px 5px 0 0;
}

ul.menu-list-drawer{
  list-style-type: none;
  margin:0;
  padding:0;
}

.menu-list-drawer li {
  display: block;
  padding: 2px 0;
}

.menu-list-drawer li div{
  display: block;
  padding: 2px 0;
  border: 1px solid #bac1ca;
  border-radius: 3px;
  cursor: pointer;
  background-color: #e6e8f1;
  &:hover {
    box-shadow:0 0 10px 0 #7b84ae;
  }
  &:active{
    background-color:#b8c2d4;
  }
}

.menu-list-drawer a{
  display: block;
  text-decoration: none;
  font-weight: bold;
  color: #000;
  font-size:1.5em;
  padding-left: 5px;

}

ul.menu-list {
  list-style-type: none;
  margin:0;
  padding: @padding-menu-list;
}

.menu-list li {
  display: inline-block;
  padding:0;
}

.menu-list a{
  text-decoration: none;
  font-weight: bold;
  color: #fff;
  font-size:1.5em;
}

.menu-list li:not(:first-child) a:before {
  content: "|";
  padding:5px;
}

.menu-list a:visited{
  color: #fff;
}

.icon-bars-block{
  padding: @padding-menu-list;
}

.icon-bars {
  padding: @padding-menu-list;
  width: 35px;
  height: 35px;
  display: block;
  background-size: 90px 150px;
  background: url('Img/bars-solid.svg') no-repeat;
  cursor: pointer;
}

.menu-close-block{
  @media @phone {
    position:relative;
    height: 42px;
    margin: 3px 0;
  }
}

.menu-close-icon{
  @media @phone {
    height: 42px;
    width: 42px;
    position:absolute;
    right:0;
    top:0;
  }
}