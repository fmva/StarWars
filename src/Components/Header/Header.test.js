import Header from './Header';
import { mount } from 'enzyme';
import React from 'react';
import { BrowserRouter, Route } from 'react-router-dom';

const menu = [
  {
    path: 'path 1',
    title: 'title 1'
  },
  {
    path: 'path 2',
    title: 'title 2'
  }
];

const onClickMenu = jest.fn();
let wrapper;

describe('Header', () => {
  beforeAll(() => {
    wrapper = mount(
      <BrowserRouter>
        <Route>
          <Header menu={menu} onClickMenu={onClickMenu} />
        </Route>
      </BrowserRouter>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('Компонент Header c входными данными', () => {
    expect(wrapper.find('div.main-size ul.menu-list a').length).toBe(2);

    expect(
      wrapper
        .find('div.main-size ul.menu-list a')
        .at(0)
        .text()
    ).toBe('title 1');
    expect(
      wrapper
        .find('div.main-size ul.menu-list a')
        .at(1)
        .text()
    ).toBe('title 2');

    expect(
      wrapper
        .find('div.main-size ul.menu-list Link')
        .at(0)
        .props().to
    ).toBe('/path 1');
    expect(
      wrapper
        .find('div.main-size ul.menu-list Link')
        .at(1)
        .props().to
    ).toBe('/path 2');
  });

  it(
    'Допустим пользователь нажимает на иконку меню в мобильной версии, ' +
      'тогда должно сработать событие onClickMenu',
    () => {
      wrapper.find('i.icon-bars').simulate('click');
      expect(onClickMenu.mock.calls.length).toBe(1);
    }
  );
});
