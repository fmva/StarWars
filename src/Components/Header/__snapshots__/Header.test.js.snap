// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Route>
      <Header
        menu={
          Array [
            Object {
              "path": "path 1",
              "title": "title 1",
            },
            Object {
              "path": "path 2",
              "title": "title 2",
            },
          ]
        }
        onClickMenu={[MockFunction]}
      >
        <header>
          <div
            className="main-size header-menu"
          >
            <ul
              className="menu-list phone-visible-none"
            >
              <li
                key="0"
              >
                <div>
                  <Link
                    replace={false}
                    to="/path 1"
                  >
                    <a
                      href="/path 1"
                      onClick={[Function]}
                    >
                      title 1
                    </a>
                  </Link>
                </div>
              </li>
              <li
                key="1"
              >
                <div>
                  <Link
                    replace={false}
                    to="/path 2"
                  >
                    <a
                      href="/path 2"
                      onClick={[Function]}
                    >
                      title 2
                    </a>
                  </Link>
                </div>
              </li>
            </ul>
            <div
              className="icon-bars-block phone-visible-block"
            >
              <i
                className="icon-bars"
                onClick={[Function]}
                title="Menu"
              />
            </div>
            <img
              alt="StarWars"
              className="sw-logo"
              src="Star-Wars-Logo.png"
            />
          </div>
          <withConfigConsumer(Drawer)
            className="list-drawer"
            closable={false}
            onClose={[Function]}
            placement="left"
            visible={false}
          >
            <Drawer
              className="list-drawer"
              closable={false}
              getPrefixCls={[Function]}
              height={256}
              level={null}
              mask={true}
              maskClosable={true}
              onClose={[Function]}
              placement="left"
              prefixCls="ant-drawer"
              renderEmpty={[Function]}
              visible={false}
              width={256}
            >
              <Drawer
                className="list-drawer"
                closable={false}
                duration=".3s"
                ease="cubic-bezier(0.78, 0.14, 0.15, 0.86)"
                firstEnter={false}
                getContainer="body"
                getPrefixCls={[Function]}
                handler={false}
                level={null}
                mask={true}
                maskClosable={true}
                maskStyle={Object {}}
                onChange={[Function]}
                onClose={[Function]}
                onHandleClick={[Function]}
                onMaskClick={[Function]}
                open={false}
                placement="left"
                prefixCls="ant-drawer"
                renderEmpty={[Function]}
                showMask={true}
                style={
                  Object {
                    "transform": undefined,
                    "zIndex": undefined,
                  }
                }
                visible={false}
                width={256}
                wrapperClassName=""
              />
            </Drawer>
          </withConfigConsumer(Drawer)>
        </header>
      </Header>
    </Route>
  </Router>
</BrowserRouter>
`;
